xarray/backends/tiledb_.py

```diff
@@ -0,0 +1,430 @@
+    through a :class:`LazilyOuterIndexedArray` object.
+    """
+
+    __slots__ = ("_array_args", "_index_converters", "shape", "dtype")
+
+    def __init__(self, attr, uri, key, timestamp, index_converters):
+        """
+        Parameters
+        ----------
+        attr : tiledb.Attr
+            The TileDB attribute being wrapped by this routine. Must be one of the
+            attributes in the array at the provided URI.
+        uri : str
+            Uniform Resoure Identifier (URI) for TileDB array. May be a path to a
+            local TileDB array or a URI for a remote resource.
+        key : Optional[str]
+            If not None, the key for accessing the TileDB array at the provided URI.
+        timestamp : Optional[int]
+            If not None, time in milliseconds to open the array at.
+        ctx : Optional[tiledb.Ctx]
+            If not None, a TileDB context manager object.
+        index_converters : Tuple[TileDBIndexConverter, ...]
+            The TileDBIndexConverters needed to convert each xarray index to its
+            corresponding TileDB coordinates.
+        """
+        if attr.isanon:
+            raise NotImplementedError(
+                "Support for anonymnous TileDB attributes has not been implemented."
+            )
+        self.dtype = attr.dtype
+        self._array_args = (uri, "r", key, timestamp, attr.name)
+        self._index_converters = index_converters
+        self.shape = tuple(converter.size for converter in index_converters)
+
+    def __getitem__(self, indexer: ExplicitIndexer):
+        xarray_indices = indexer.tuple
+        if len(xarray_indices) != len(self._index_converters):
+            raise ValueError(
+                f"key of length {len(xarray_indices)} cannot be used for a TileDB array"
+                f" of length {len(self._index_converters)}"
+            )
+        shape = tuple(
+            len(range(converter.size)[index] if isinstance(index, slice) else index)
+            for index, converter in zip(xarray_indices, self._index_converters)
+            if not isinstance(index, integer_types)
+        )
+        # TileDB multi_index does not except empty arrays/slices. If a dimension is
+        # length zero, return an empty numpy array of the correct length.
+        if 0 in shape:
+            return np.zeros(shape)
+        tiledb_indices = tuple(
+            converter[index]
+            for index, converter in zip(xarray_indices, self._index_converters)
+        )
+        with tiledb.DenseArray(*self._array_args) as array:
+            result = array.multi_index[tiledb_indices][self._array_args[-1]]
+        # Note: TileDB multi_index returns the same number of dimensions as the initial
+        # array. To match the expected xarray output, we need to reshape the result to
+        # remove any dimensions corresponding to scalar-valued input.
+        return result.reshape(shape)
+
+
+class TileDBDataStore(AbstractDataStore):
```

Comment:
I'm not sure if this is a misconception on my part, but I think `DataStore` and `AbstractDataStore` are part of the old API and are replaced by `BackendEntrypoint`. Any comments on that, @alexamici, @aurghs?